In [110]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns


In [111]:
df = pd.read_csv("energydata_complete.csv")

Question 17

From the dataset, fit a linear model on the relationship between the temperature in the living room in Celsius (x = T2) and the temperature outside the building (y = T6). What is the Root Mean Squared error in three D.P?







In [113]:
#sns.regplot(x=df["T2"], y=["T6"])

### question 18

Remove the following columns: [“date”, “lights”]. The target variable is “Appliances”. Use a 70-30 train-test set split with a  random state of 42 (for reproducibility). Normalize the dataset using the MinMaxScaler (Hint: Use the MinMaxScaler fit_transform and transform methods on the train and test set respectively). Run a multiple linear regression using the training set. Answer the following questions:

What is the Mean Absolute Error (in three decimal places) for the  training set?







In [69]:
df2 = df.drop(["date", "lights"], axis = 1)

In [70]:
y = df2["Appliances"]
x = df2.drop(["Appliances"], axis = 1)
from sklearn.model_selection import train_test_split
x_train, x_test, y_train, y_test = train_test_split( x, y, test_size = 0.3, 
                                                    random_state = 42)

In [71]:
from sklearn.preprocessing import MinMaxScaler
scaler = MinMaxScaler()
# x_train, y_train, x_test, y_test = pd.DataFrame(scaler.fit_transform([x_train, y_train, x_test, y_test]))
x_train_normalized = scaler.fit_transform(x_train)
x_test_normalized = scaler.transform(x_test)

In [72]:
from sklearn.linear_model import LinearRegression
linear_model = LinearRegression()
# fit the model to the training dataset
linear_model.fit(x_train, y_train)

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=None, normalize=False)

In [80]:
predict_on_train = linear_model.predict(x_train)

In [82]:
from sklearn.metrics import mean_absolute_error
round(mean_absolute_error(y_train, predict_on_train),3)


53.742

### Question 19

What is the Root Mean Squared Error (in three decimal places) for the training set?

In [83]:
import numpy as np
from sklearn.metrics import mean_squared_error
round(np.sqrt(mean_squared_error(y_train, predict_on_train)),3)


95.216

### Question 20

What is the Mean Absolute Error (in three decimal places) for test set?

In [85]:
predict_on_test = linear_model.predict(x_test)
from sklearn.metrics import mean_absolute_error
round(mean_absolute_error(y_test, predict_on_test),3)


53.643

### Question 21

What is the Root Mean Squared Error (in three decimal places) for test set?



In [86]:
round(np.sqrt(mean_squared_error(y_test, predict_on_test)),3)

93.64

* RMSE of train = 95.216
* RMSE of test = 93.64
* RMSE of test < RMSE of train

### Question 23

Train a ridge regression model with default parameters. Is there any change to the root mean squared error (RMSE) when evaluated on the test set?

In [87]:
from sklearn.linear_model import Ridge
ridge_regression = Ridge()
ridge_regression.fit(x_train, y_train)

Ridge(alpha=1.0, copy_X=True, fit_intercept=True, max_iter=None,
      normalize=False, random_state=None, solver='auto', tol=0.001)

In [89]:
ridge_on_test = ridge_regression.predict(x_test)

In [90]:
round(np.sqrt(mean_squared_error(y_test, ridge_on_test)),3)

93.641

* RMSE with Ridge regression = 93.641
* RMSE with linear model = 93.64

### Question 24

Train a lasso regression model with default value and obtain the new feature weights with it. How many of the features have non-zero feature weights?

In [92]:
from sklearn.linear_model import Lasso
lasso_regression = Lasso()
lasso_regression.fit(x_train, y_train)



Lasso(alpha=1.0, copy_X=True, fit_intercept=True, max_iter=1000,
      normalize=False, positive=False, precompute=False, random_state=None,
      selection='cyclic', tol=0.0001, warm_start=False)

### Question 25

What is the new RMSE with the Lasso Regression on the test set?

In [94]:
test_set_with_lasso = lasso_regression.predict(x_train)

In [102]:
round(np.sqrt(mean_squared_error(y_test, test_set_with_lasso),3))